File: leetcode/Go/code/206.反转链表/206.反转链表.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList1(head *ListNode) *ListNode {
	var pre *ListNode = nil
	var cur *ListNode = head

	for cur != nil {
		var tmp = cur.Next
		cur.Next = pre
		pre = cur
		cur = tmp
	}

	return pre
}

func reverseListRecursion(head *ListNode) *ListNode {
	// 1. 递归终止条件
	if head == nil || head.Next == nil {
		return head
	}
	var p = reverseListRecursion(head.Next)
	head.Next.Next = head
	head.Next = nil
	return p
}

File: leetcode/Python/code/206.反转链表/206.反转链表.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr, prev = head, None

        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp

        return prev

    def reverseListRecursion(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head

        r = self.reverseListRecursion(head.next)
        head.next.next = head
        head.next = None
        return r
